addons/website_sale_loyalty/models/sale_order.py

```diff
@@ -0,0 +1,330 @@
+            product = reward.discount_product_id
+            if not product:
+                return
+            taxes = product.taxes_id.filtered(lambda t: t.company_id.id == self.company_id.id)
+            if self.fiscal_position_id:
+                taxes = self.fiscal_position_id.map_tax(taxes)
+            order_line = self.env['sale.order.line'].create({
+                'product_id': product.id,
+                'loyalty_reward_id': reward.id,
+                'is_main_loyalty_reward': True,
+                'product_uom_qty': 1,
+                'product_uom': product.uom_id.id,
+                'order_id': self.id,
+                'price_unit': 0,  # computed by _cart_adjust_rewards
+                'tax_id': [fields.Command.link(tax.id) for tax in taxes],
+                'sequence': REWARD_SALE_ORDER_LINE_SEQUENCE,
+            })
+        self._cart_adjust_rewards()
+        website = self.env['website'].get_current_website()
+        self.recompute_loyalty_points(website.loyalty_id.id)
+
+    def _get_won_points(self, loyalty):
+        """The total of points won, excluding the points spent on rewards"""
+        if self.website_id.is_public_user() or not loyalty:
+            return 0
+        total_points = 0
+        for line in self.order_line:
+            if line.loyalty_reward_id:  # Reward products are ignored
+                continue
+
+            line_points = 0
+            for rule in loyalty.rule_ids:
+                rule_points = 0
+                if rule.is_product_valid(line.product_id):
+                    rule_points += rule.points_quantity * line.product_uom_qty
+                    rule_points += rule.points_currency * line.price_total
+                if rule_points > line_points:
+                    line_points = rule_points
+
+            total_points += line_points
+
+        total_points += self.amount_total * loyalty.points
+        return tools.float_round(total_points, 0, rounding_method='HALF-UP')
+
+    def _get_spent_points(self):
+        """The total number of points spent on rewards"""
+        if not self.partner_id:
+            return 0
+        points = 0
+        for line in self.order_line:
+            if line.loyalty_reward_id and line.is_main_loyalty_reward:
+                line_points = line.product_uom_qty * line.loyalty_reward_id.point_cost
+                points += tools.float_round(line_points, 0, rounding_method='HALF-UP')
+        return points
+
+    def recompute_loyalty_points(self, loyalty_id):
+        loyalty = self.env['loyalty.program'].browse(loyalty_id) if loyalty_id else None
+        for order in self:
+            if loyalty:
+                order.won_loyalty_points = order._get_won_points(loyalty)
+                order.spent_loyalty_points = order._get_spent_points()
+            else:
+                order.won_loyalty_points = 0
+                order.spent_loyalty_points = 0
+
+    def get_portal_loyalty_url(self):
```

Comment:
Why not, I personally think it is a bit overkill to have this 'hack' here and on the /my/loyalty/<id> controller to just have /my/loyalty in the URL.
Without this, it would work exactly the same except the URL right?
We would still have /my/loyalty which list order, then you land on /my/order/id instead of /my/loyalty/id, but the breadcrumb still show "Loyalty" and redirect correctly to /my/loyalty if I am not mistaking?

I'd think the less code the better in this case, especially since they are kind of hacks, but mater of taste :-)

Reply:
The breadcrumb would be "Sale Order" instead of "Loyalty", because it is running through that route that sets the `page_name` attribute used in the template to differentiate them.
I do not see how to recognize that we reach the page from Loyalty if the whole request is identical.

Reply:
Isn't there any other solution? An url param inside the /my/loyalty templates so the /my/order/4 knows about it?
Really not a big fan of this, if `orders` is one day changed in the route, there is 100% chance this method won't be updated and will simply stop working. But I can live with it ;-)